flow/Arena.h

```diff
@@ -783,7 +783,7 @@ struct VectorRefPreserializer<T, VecSerStrategy::String> {
 };
 
 template <class T, VecSerStrategy SerStrategy = VecSerStrategy::FlatBuffers>
-class VectorRef : public ComposedIdentifier<T, 0x8>, public VectorRefPreserializer<T, SerStrategy> {
+class VectorRef : public ComposedIdentifier<T, 3>, public VectorRefPreserializer<T, SerStrategy> {
```

Comment:
I'm not familiar with this area. 
Can you explain why 0x8 is changed to 3? From the change above, I thought you just changed the value from base 2 to base 10. 

Maybe another question is:
how to decide which value should the template `B` be? (In this case, why it is 3 instead of 4 or 5?)

Reply:
In the old composed file identifier implementation, if a type had multiple composed identifiers, those were both represented in the first 8 bits using a bitwise or. Thus, the first 8 bits of the `VectorRef<ReplyPromise<T>>` would be `0x8 | 0x2`. This had several problems because `ComposedType1<ComposedType2<T>>` would have the same identifier at `ComposedType2<ComposedType1<T>>` and `ComposedType<ComposedType<T>>` had the same file identifier as `ComposedType<T>`. So now, in the new implementation, the first 4 bits are reserved for the outer layer of composition, and the next 4 bits are used for the inner layer of composition. To save bits, we changed the values of `B` from `1,2,4,8...` to `0,1,2,3...`, because there's no need to worry about the result of applying a bitwise or to two values of `B`.

All of the `B`s have to be unique for different composed types, and they all have to be in the range `[0,15]`. Beyond that, it doesn't matter what `B` is, anyone adding a new composed type will have to manually make sure its `B` is unique, but this is a very rare change (we only have 5 distinct composed types so far). There is a static assertion to enforce that `B` is always in the range `[0,15]`.